### ISSUE TITLE
FEFieldFunction is not thread-safe (because GridTools is not threadsafe)

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#8516]
@mjayadharan showed me a piece of code that crashes in parallel. I can't say I completely understand why it does, but in looking into it, I think that the following is nonetheless a problem:

When one uses `FEFieldFunction` as an argument to places such as `MatrixTools::create_mass_matrix()` or a variety of other places that build stuff in parallel, one might get into the `vector_value_list` function from multiple threads at the same time:
```
  void
  FEFieldFunction<dim, DoFHandlerType, VectorType>::vector_value_list(
    const std::vector<Point<dim>> &                       points,
    std::vector<Vector<typename VectorType::value_type>> &values) const
  {
    Assert(...);

    std::vector<typename DoFHandlerType::active_cell_iterator> cells;
    std::vector<std::vector<Point<dim>>>                       qpoints;
    std::vector<std::vector<unsigned int>>                     maps;

    const unsigned int n_cells =
      compute_point_locations(points, cells, qpoints, maps);
```
So we can also compute `compute_point_locations()` multiple times in parallel:
```
  unsigned int
  FEFieldFunction<dim, DoFHandlerType, VectorType>::compute_point_locations(
    const std::vector<Point<dim>> &                             points,
    std::vector<typename DoFHandlerType::active_cell_iterator> &cells,
    std::vector<std::vector<Point<dim>>> &                      qpoints,
    std::vector<std::vector<unsigned int>> &                    maps) const
  {
    // Calling the GridTools routine and preparing output
    auto cell_qpoint_map =
      GridTools::compute_point_locations(cache, points, cell_hint.get());
  ...
```

From there, through a couple more places, we end up in the `GridTools::Cache` class, possibly again multiple times in parallel:
```
  Cache<dim, spacedim>::get_cell_bounding_boxes_rtree() const
  {
    if (update_flags & update_cell_bounding_boxes_rtree)
      {
        std::vector<std::pair<
          BoundingBox<spacedim>,
          typename Triangulation<dim, spacedim>::active_cell_iterator>>
                     boxes(tria->n_active_cells());
        unsigned int i = 0;
        for (const auto &cell : tria->active_cell_iterators())
          boxes[i++] = std::make_pair(mapping->get_bounding_box(cell), cell);

        cell_bounding_boxes_rtree = pack_rtree(boxes);
        ...
```
The issue is the assignment in this last line here, which is to a member variable of the `Cache` class. As part of the assignment, previous content is destroyed and new content is assigned. This can't work if it happens from multiple threads at the same time.

In other words, either the `FEFieldFunction` or the `GridTools::Cache` class will have to learn how to be thread-safe.

@luca-heltai -- FYI.

**Comment by bangerth** [comment]
I think I know what needs to be done here.
